staging/src/k8s.io/endpointslice/trafficdist/trafficdist.go

```diff
@@ -101,43 +109,61 @@ func (defaultHeuristic) update(slice *discoveryv1.EndpointSlice) {
 	}
 }
 
-// preferCloseHeuristic adds
+// preferCloseHeuristic implements PreferSameZone/PreferClose and PreferSameNode
 type preferCloseHeuristic struct {
+	generateNodeHints bool
 }
 
 // needsUpdate returns true if any ready endpoint in the slice has a
 // missing or incorrect hint.
-func (preferCloseHeuristic) needsUpdate(slice *discoveryv1.EndpointSlice) bool {
+func (h preferCloseHeuristic) needsUpdate(slice *discoveryv1.EndpointSlice) bool {
 	if slice == nil {
 		return false
 	}
 	for _, endpoint := range slice.Endpoints {
 		if !endpointReady(endpoint) {
 			continue
 		}
-		var zone string
-		if endpoint.Zone != nil {
-			zone = *endpoint.Zone
+
+		// We should have a zone hint if and only if the endpoints have associated zones.
+		wantZoneHint := endpoint.Zone != nil
+		hasZoneHint := endpoint.Hints != nil && endpoint.Zone != nil && len(endpoint.Hints.ForZones) == 1 && endpoint.Hints.ForZones[0].Name == *endpoint.Zone
+		if wantZoneHint != hasZoneHint {
+			return true
 		}
 
-		if endpoint.Hints == nil || len(endpoint.Hints.ForZones) != 1 || endpoint.Hints.ForZones[0].Name != zone {
+		wantNodeHint := endpoint.NodeName != nil && h.generateNodeHints
```

Comment:
I'm wondering about the case where generateNodeHInts=false (maybe someone switch trafficDist field) and the nodeHint is currently incorrect. In such a case, the current behaviour seems to not remove the hint. I know our own code will never lead to such a state (unless there's a bug somewhere). Wondering if that's worth accounting for or not?

Reply:
the code is supposed to deal with all cases, but you're right, it doesn't